server/resp/src/main/java/org/infinispan/server/resp/hll/internal/CompactSet.java

```diff
@@ -0,0 +1,291 @@
+ * </p>
+ *
+ * <p>
+ * In C, we could allocate the memory space and manipulate it as we want, but here in Java, we have to do some bit
+ * twiddling magic to make the registers fit. In other words, each position in the array uses long with 64 bits, whereas
+ * one register uses only 6 bits. Multiple registers are within a single array position or spread across two positions.
+ * Some manipulation is necessary to retrieve the correct register value from the array.
+ * </p>
+ *
+ * <p>
+ * The basic idea is that with a 64-bit hash, the first P bits identify the bucket, and the remaining Q bits we use to
+ * count the trailing zeroes, and in case all bits are 0, the count is Q + 1. The next step is to update the bucket value
+ * if it is smaller than the current count. From the bucket number and some bit-shifting, we identify the register.
+ * </p>
+ *
+ * <p><b>Credits</b></p>
+ * This implementation takes inspiration from the Redis implementation [1], which is based [2]. The bit-shifting
+ * manipulation has some basis in [3], which implements the original algorithm without optimizations.
+ *
+ * <p>
+ * Our cardinality estimation uses the optimizations proposed in [2]. And this is the reason we keep an additional int
+ * array. We use this approach as it is the same as Redis uses. This algorithm does not require a bias correction as the
+ * Google implementation or the [3]. That is, it does not need empirical data to correct estimations.
+ * </p>
+ *
+ * @see <a href="https://github.com/redis/redis/blob/unstable/src/hyperloglog.c">[1] Redis HyperLogLog implementation.</a>
+ * @see <a href="https://arxiv.org/pdf/1702.01284.pdf">[2] New cardinality estimation algorithms for HyperLogLog sketches</a>
+ * @see <a href="https://github.com/aggregateknowledge/java-hll">[3] Aggregate Knowledge HyperLogLog implementation.</a>
+ * @since 15.0
+ * @author José Bolina
+ */
+@ProtoTypeId(ProtoStreamTypeIds.RESP_HYPER_LOG_LOG_COMPACT)
+public class CompactSet implements HLLRepresentation {
+
+   // The size of the actual register we use to store the count of zeroes.
+   // The highest number we can have is `Util.HLL_Q + 1`, which is representable in 6 bits.
+   static final long REGISTER_WIDTH = 6;
+
+   // The maximum value a register can hold with the given size.
+   static final long MAX_REGISTER_ENTRY = 1 << REGISTER_WIDTH;
+
+   // Mask to identify the correct register in the store position.
+   static final long SINGLE_REGISTER_MASK = MAX_REGISTER_ENTRY - 1;
+
+   // The total store size. This takes into account the splitted registers and the multiple registers per array position.
+   // In total, this will occupy roughly 12Kb.
+   static final int STORE_SIZE = (int) (((REGISTER_WIDTH * Util.HLL_REGISTERS) + SINGLE_REGISTER_MASK) >>> REGISTER_WIDTH);
+
+   private final long[] store;
+   private final int[] multiplicity;
+   private byte minimum;
+
+   public CompactSet() {
+      this.store = new long[STORE_SIZE];
```

Comment:
Note that the other approach here would be `byte[Util.HLL_REGISTERS]`, and we access it directly. That would create a slightly bigger object (~16Kb against the ~12Kb) but simplify the code a ton.